router/h2/src/main/scala/com/twitter/finagle/buoyant/h2/H2FailureAccrualFactory.scala

```diff
@@ -0,0 +1,276 @@
+
+  def module: Stackable[ServiceFactory[Request, Response]] =
+    new Stack.ModuleParams[ServiceFactory[Request, Response]] {
+      val role: Role = FailureAccrualFactory.role
+      val description: String = "Backoff from hosts that we cannot successfully make requests to"
+      override def parameters: Seq[Stack.Param[_]] = Seq(
+        implicitly[Stack.Param[fParam.Stats]],
+        implicitly[Stack.Param[FailureAccrualFactory.Param]],
+        implicitly[Stack.Param[fParam.Timer]],
+        implicitly[Stack.Param[fParam.Label]],
+        implicitly[Stack.Param[fParam.Logger]],
+        implicitly[Stack.Param[Transporter.EndpointAddr]]
+      )
+
+      def make(params: Params, next: ServiceFactory[Request, Response]): ServiceFactory[Request, Response] = {
+        params[FailureAccrualFactory.Param] match {
+          case FailureAccrualFactory.Param.Configured(policy) =>
+            val timer = params[fParam.Timer].timer
+            val statsReceiver = params[fParam.Stats].statsReceiver
+
+            // extract some info useful for logging
+            val logger = params[fParam.Logger].log
+            val endpoint = params[Transporter.EndpointAddr].addr
+            val label = params[fParam.Label].label
+
+            new H2FailureAccrualFactory(
+              underlying = next,
+              policy = policy(),
+              timer = timer,
+              statsReceiver = statsReceiver.scope("failure_accrual")
+            ) {
+              override def didMarkDead(): Unit = {
+                logger.log(
+                  Level.INFO,
+                  s"""FailureAccrualFactory marking connection to "$label" as dead. """ +
+                    s"""Remote Address: $endpoint"""
+                )
+                super.didMarkDead()
+              }
+            }
+
+          case FailureAccrualFactory.Param.Replaced(f) =>
+            f(params[fParam.Timer].timer).andThen(next)
+
+          case FailureAccrualFactory.Param.Disabled =>
+            next
+        }
+      }
+    }
+
+  // The FailureAccrualFactory transitions between Alive, Dead, ProbeOpen,
+  // and ProbeClosed. The factory starts in the Alive state. After numFailures
+  // failures, the factory transitions to Dead. When it is revived,
+  // it transitions to ProbeOpen. After a request is received,
+  // it transitions to ProbeClosed and cannot accept any further requests until
+  // the initial request is satisfied. If the request is successful, it
+  // transitions back to Alive, otherwise Dead.
+  //
+  // The transitions can be visualized using the state diagram:
+  //
+  // ,<-----------.
+  // Alive        |
+  // |  ,---ProbeClosed
+  // ∨  ∨         ^
+  // Dead         |
+  //  `---> ProbeOpen
+  protected[finagle] sealed trait State
+  protected[finagle] object Alive extends State
+  protected[finagle] object Dead extends State
+  protected[finagle] object ProbeOpen extends State
+  protected[finagle] object ProbeClosed extends State
```

Comment:
Do you think we should just reuse the stat definitions in `FailureAccrualFactory`?